File: examples/python/fenwick_tree_examples.py
class FenwickTree:
    def __init__(self, n):
        self.n = n
        self.tree = [0] * (n + 1)
    
    def update(self, i, delta):
        while i <= self.n:
            self.tree[i] += delta
            i += i & (-i)
    
    def query(self, i):
        result = 0
        while i > 0:
            result += self.tree[i]
            i -= i & (-i)
        return result
    
    def range_query(self, l, r):
        return self.query(r) - self.query(l - 1)

class NumArray:
    def __init__(self, nums):
        self.nums = nums[:]
        self.bit = FenwickTree(len(nums))
        for i, num in enumerate(nums):
            self.bit.update(i + 1, num)
    
    def update(self, index, val):
        delta = val - self.nums[index]
        self.nums[index] = val
        self.bit.update(index + 1, delta)
    
    def sumRange(self, left, right):
        return self.bit.range_query(left + 1, right + 1)

# 1. Count Smaller
def count_smaller(nums):
    sorted_nums = sorted(set(nums))
    coord_map = {v: i + 1 for i, v in enumerate(sorted_nums)}
    
    bit = FenwickTree(len(sorted_nums))
    result = []
    
    for num in reversed(nums):
        idx = coord_map[num]
        count = bit.query(idx - 1)
        result.append(count)
        bit.update(idx, 1)
    
    return result[::-1]

# 2. Reverse Pairs
def reverse_pairs(nums):
    sorted_nums = sorted(set(nums + [2 * x for x in nums]))
    coord_map = {v: i + 1 for i, v in enumerate(sorted_nums)}
    
    bit = FenwickTree(len(sorted_nums))
    count = 0
    
    for num in reversed(nums):
        count += bit.query(coord_map[2 * num] - 1)
        bit.update(coord_map[num], 1)
    
    return count

class RangeSumQuery2D:
    def __init__(self, matrix):
        if not matrix or not matrix[0]:
            return
        self.m, self.n = len(matrix), len(matrix[0])
        self.matrix = [[0] * self.n for _ in range(self.m)]
        self.bit = [[0] * (self.n + 1) for _ in range(self.m + 1)]
        
        for i in range(self.m):
            for j in range(self.n):
                self.update(i, j, matrix[i][j])
    
    def update(self, row, col, val):
        delta = val - self.matrix[row][col]
        self.matrix[row][col] = val
        
        i = row + 1
        while i <= self.m:
            j = col + 1
            while j <= self.n:
                self.bit[i][j] += delta
                j += j & (-j)
            i += i & (-i)
    
    def query(self, row, col):
        if row < 0 or col < 0:
            return 0
        result = 0
        i = row + 1
        while i > 0:
            j = col + 1
            while j > 0:
                if i < len(self.bit) and j < len(self.bit[i]):
                    result += self.bit[i][j]
                j -= j & (-j)
            i -= i & (-i)
        return result
    
    def sumRegion(self, row1, col1, row2, col2):
        return (self.query(row2, col2) - self.query(row1 - 1, col2) - 
                self.query(row2, col1 - 1) + self.query(row1 - 1, col1 - 1))

# 3. Count Range Sum
def count_range_sum(nums, lower, upper):
    prefix_sums = [0]
    for num in nums:
        prefix_sums.append(prefix_sums[-1] + num)
    
    sorted_sums = sorted(set(prefix_sums))
    coord_map = {v: i + 1 for i, v in enumerate(sorted_sums)}
    
    bit = FenwickTree(len(sorted_sums))
    count = 0
    
    for prefix_sum in prefix_sums:
        left_idx = bisect_left(sorted_sums, prefix_sum - upper)
        right_idx = bisect_right(sorted_sums, prefix_sum - lower) - 1
        
        if left_idx <= right_idx:
            count += bit.range_query(left_idx + 1, right_idx + 1)
        
        bit.update(coord_map[prefix_sum], 1)
    
    return count

# 4. Bisect Left
def bisect_left(arr, x):
    left, right = 0, len(arr)
    while left < right:
        mid = (left + right) // 2
        if arr[mid] < x:
            left = mid + 1
        else:
            right = mid
    return left

# 5. Bisect Right
def bisect_right(arr, x):
    left, right = 0, len(arr)
    while left < right:
        mid = (left + right) // 2
        if arr[mid] <= x:
            left = mid + 1
        else:
            right = mid
    return left

class CountIntervals:
    def __init__(self):
        self.intervals = []
        self.bit = FenwickTree(10**6)
    
    def add(self, left, right):
        self.intervals.append((left, right))
        for i in range(left, right + 1):
            self.bit.update(i, 1)
    
    def count(self):
        return len(set().union(*[set(range(l, r + 1)) for l, r in self.intervals]))

# 6. Create Sorted Array
def create_sorted_array(instructions):
    MOD = 10**9 + 7
    sorted_nums = sorted(set(instructions))
    coord_map = {v: i + 1 for i, v in enumerate(sorted_nums)}
    
    bit = FenwickTree(len(sorted_nums))
    cost = 0
    
    for num in instructions:
        idx = coord_map[num]
        smaller = bit.query(idx - 1)
        larger = bit.query(len(sorted_nums)) - bit.query(idx)
        cost = (cost + min(smaller, larger)) % MOD
        bit.update(idx, 1)
    
    return cost

class FenwickTree2D:
    def __init__(self, m, n):
        self.m, self.n = m, n
        self.tree = [[0] * (n + 1) for _ in range(m + 1)]
    
    def update(self, x, y, delta):
        i = x + 1
        while i <= self.m:
            j = y + 1
            while j <= self.n:
                self.tree[i][j] += delta
                j += j & (-j)
            i += i & (-i)
    
    def query(self, x, y):
        result = 0
        i = x + 1
        while i > 0:
            j = y + 1
            while j > 0:
                result += self.tree[i][j]
                j -= j & (-j)
            i -= i & (-i)
        return result

# 7. Max Sum Rectangle
def max_sum_rectangle(matrix, k):
    m, n = len(matrix), len(matrix[0])
    max_sum = float('-inf')
    
    for top in range(m):
        temp = [0] * n
        for bottom in range(top, m):
            for j in range(n):
                temp[j] += matrix[bottom][j]
            
            # Find max subarray sum <= k
            prefix_sums = [0]
            for val in temp:
                prefix_sums.append(prefix_sums[-1] + val)
            
            for i in range(len(prefix_sums)):
                for j in range(i + 1, len(prefix_sums)):
                    current_sum = prefix_sums[j] - prefix_sums[i]
                    if current_sum <= k:
                        max_sum = max(max_sum, current_sum)
    
    return max_sum

# 8. Range Add Query
def range_add_query(length, updates):
    bit = FenwickTree(length)
    
    for left, right, val in updates:
        bit.update(left + 1, val)
        if right + 2 <= length:
            bit.update(right + 2, -val)
    
    result = []
    for i in range(length):
        result.append(bit.query(i + 1))
    return result

# 9. Count Inversions
def count_inversions(nums):
    sorted_nums = sorted(set(nums))
    coord_map = {v: i + 1 for i, v in enumerate(sorted_nums)}
    
    bit = FenwickTree(len(sorted_nums))
    count = 0
    
    for num in reversed(nums):
        idx = coord_map[num]
        count += bit.query(idx - 1)
        bit.update(idx, 1)
    
    return count

# 10. Run Examples
def run_examples():
    print("=== Fenwick Tree Examples ===")
    
    # Test 1: Range Sum Query
    nums = [1, 3, 5, 7, 9, 11]
    num_array = NumArray(nums)
    print(f"Sum Range [1,3]: {num_array.sumRange(1, 3)}")
    num_array.update(1, 10)
    print(f"Sum Range [1,3] after update: {num_array.sumRange(1, 3)}")
    
    # Test 2: Count Smaller Numbers
    nums = [5, 2, 6, 1]
    print(f"Count Smaller: {count_smaller(nums)}")
    
    # Test 3: Reverse Pairs
    nums = [1, 3, 2, 3, 1]
    print(f"Reverse Pairs: {reverse_pairs(nums)}")
    
    # Test 4: 2D Range Sum Query
    matrix = [[3, 0, 1, 4, 2], [5, 6, 3, 2, 1], [1, 2, 0, 1, 5]]
    range_sum_2d = RangeSumQuery2D(matrix)
    print(f"2D Sum Region [1,1,2,2]: {range_sum_2d.sumRegion(1, 1, 2, 2)}")
    
    # Test 5: Create Sorted Array
    instructions = [1, 5, 6, 2]
    print(f"Create Sorted Array Cost: {create_sorted_array(instructions)}")
    
    # Test 6: Range Add Query
    updates = [[1, 3, 2], [2, 4, 3], [0, 2, -2]]
    print(f"Range Add Query: {range_add_query(5, updates)}")
    
    # Test 7: Count Range Sum
    nums = [-2, 5, -1]
    lower, upper = -2, 2
    print(f"Count Range Sum: {count_range_sum(nums, lower, upper)}")
    
    # Test 8: 2D Fenwick Tree
    bit_2d = FenwickTree2D(3, 3)
    bit_2d.update(1, 1, 5)
    bit_2d.update(2, 2, 3)
    print(f"2D Fenwick Query (2,2): {bit_2d.query(2, 2)}")
    
    # Test 9: Max Sum Rectangle
    matrix = [[1, 0, 1], [0, -2, 3]]
    k = 2
    print(f"Max Sum Rectangle <= {k}: {max_sum_rectangle(matrix, k)}")
    
    # Test 10: Count Intervals
    counter = CountIntervals()
    counter.add(2, 3)
    counter.add(4, 5)
    counter.add(6, 7)
    print(f"Count Intervals: {counter.count()}")

if __name__ == "__main__":
    run_examples()

File: examples/java/FenwickTreeExamples.java
import java.util.*;

public class FenwickTreeExamples {
    
    static class FenwickTree {
        int[] tree;
        int n;
        
        public FenwickTree(int n) {
            this.n = n;
            this.tree = new int[n + 1];
        }
        
        public void update(int i, int delta) {
            for (++i; i <= n; i += i & -i) tree[i] += delta;
        }
        
        public int query(int i) {
            int sum = 0;
            for (++i; i > 0; i -= i & -i) sum += tree[i];
            return sum;
        }
        
        public int rangeQuery(int l, int r) {
            return query(r) - (l > 0 ? query(l - 1) : 0);
        }
    }
    
    // 1. Range Sum Query - Mutable
    static class NumArray {
        FenwickTree ft;
        int[] nums;
        
        public NumArray(int[] nums) {
            this.nums = nums.clone();
            ft = new FenwickTree(nums.length);
            for (int i = 0; i < nums.length; i++) {
                ft.update(i, nums[i]);
            }
        }
        
        public void update(int index, int val) {
            ft.update(index, val - nums[index]);
            nums[index] = val;
        }
        
        public int sumRange(int left, int right) {
            return ft.rangeQuery(left, right);
        }
    }
    
    // 2. Count of Smaller Numbers After Self
    public List<Integer> countSmaller(int[] nums) {
        int n = nums.length;
        Integer[] result = new Integer[n];
        int[] sorted = nums.clone();
        Arrays.sort(sorted);
        
        Map<Integer, Integer> ranks = new HashMap<>();
        int rank = 0;
        for (int num : sorted) {
            if (!ranks.containsKey(num)) ranks.put(num, rank++);
        }
        
        FenwickTree ft = new FenwickTree(rank);
        for (int i = n - 1; i >= 0; i--) {
            int r = ranks.get(nums[i]);
            result[i] = r > 0 ? ft.query(r - 1) : 0;
            ft.update(r, 1);
        }
        
        return Arrays.asList(result);
    }
    
    // 3. Count of Range Sum
    public int countRangeSum(int[] nums, int lower, int upper) {
        long[] prefixSums = new long[nums.length + 1];
        for (int i = 0; i < nums.length; i++) {
            prefixSums[i + 1] = prefixSums[i] + nums[i];
        }
        
        Set<Long> allSums = new HashSet<>();
        for (long sum : prefixSums) {
            allSums.add(sum);
            allSums.add(sum - lower);
            allSums.add(sum - upper);
        }
        
        Long[] sorted = allSums.toArray(new Long[0]);
        Arrays.sort(sorted);
        Map<Long, Integer> ranks = new HashMap<>();
        for (int i = 0; i < sorted.length; i++) {
            ranks.put(sorted[i], i);
        }
        
        FenwickTree ft = new FenwickTree(sorted.length);
        int count = 0;
        
        for (long sum : prefixSums) {
            int left = ranks.get(sum - upper);
            int right = ranks.get(sum - lower);
            count += ft.rangeQuery(left, right);
            ft.update(ranks.get(sum), 1);
        }
        
        return count;
    }
    
    // 4. Reverse Pairs
    public int reversePairs(int[] nums) {
        return mergeSort(nums, 0, nums.length - 1);
    }
    
    private int mergeSort(int[] nums, int left, int right) {
        if (left >= right) return 0;
        int mid = left + (right - left) / 2;
        int count = mergeSort(nums, left, mid) + mergeSort(nums, mid + 1, right);
        
        int j = mid + 1;
        for (int i = left; i <= mid; i++) {
            while (j <= right && nums[i] > 2L * nums[j]) j++;
            count += j - (mid + 1);
        }
        
        Arrays.sort(nums, left, right + 1);
        return count;
    }
    
    // 5. Count Number of Teams
    public int numTeamsInArray(int[] rating) {
        int n = rating.length;
        int count = 0;
        
        for (int j = 1; j < n - 1; j++) {
            int leftSmaller = 0, leftLarger = 0;
            int rightSmaller = 0, rightLarger = 0;
            
            for (int i = 0; i < j; i++) {
                if (rating[i] < rating[j]) leftSmaller++;
                else if (rating[i] > rating[j]) leftLarger++;
            }
            
            for (int k = j + 1; k < n; k++) {
                if (rating[k] < rating[j]) rightSmaller++;
                else if (rating[k] > rating[j]) rightLarger++;
            }
            
            count += leftSmaller * rightLarger + leftLarger * rightSmaller;
        }
        
        return count;
    }
    
    // 6. Queries on a Permutation With Key
    public int[] processQueries(int[] queries, int m) {
        FenwickTree ft = new FenwickTree(2 * m);
        int[] pos = new int[m + 1];
        
        for (int i = 1; i <= m; i++) {
            pos[i] = m + i;
            ft.update(m + i, 1);
        }
        
        int[] result = new int[queries.length];
        int nextPos = m;
        
        for (int i = 0; i < queries.length; i++) {
            int val = queries[i];
            result[i] = ft.query(pos[val] - 1);
            ft.update(pos[val], -1);
            pos[val] = nextPos--;
            ft.update(pos[val], 1);
        }
        
        return result;
    }
    
    // 7. Create Sorted Array through Instructions
    public int createSortedArray(int[] instructions) {
        int MOD = 1_000_000_007;
        int maxVal = Arrays.stream(instructions).max().getAsInt();
        FenwickTree ft = new FenwickTree(maxVal);
        
        long cost = 0;
        for (int i = 0; i < instructions.length; i++) {
            int val = instructions[i];
            int smaller = val > 1 ? ft.query(val - 2) : 0;
            int larger = ft.query(maxVal - 1) - ft.query(val - 1);
            cost = (cost + Math.min(smaller, larger)) % MOD;
            ft.update(val - 1, 1);
        }
        
        return (int)cost;
    }
    
    // 8. Range Add Query
    public int[] rangeAddQuery(int length, int[][] updates) {
        FenwickTree ft = new FenwickTree(length);
        
        for (int[] update : updates) {
            int left = update[0], right = update[1], val = update[2];
            ft.update(left, val);
            if (right + 1 < length) ft.update(right + 1, -val);
        }
        
        int[] result = new int[length];
        for (int i = 0; i < length; i++) {
            result[i] = ft.query(i);
        }
        return result;
    }
    
    // 9. Count Inversions
    public int countInversions(int[] nums) {
        int[] sorted = nums.clone();
        Arrays.sort(sorted);
        
        Map<Integer, Integer> ranks = new HashMap<>();
        int rank = 0;
        for (int num : sorted) {
            if (!ranks.containsKey(num)) ranks.put(num, rank++);
        }
        
        FenwickTree ft = new FenwickTree(rank);
        int count = 0;
        
        for (int i = nums.length - 1; i >= 0; i--) {
            int r = ranks.get(nums[i]);
            count += r > 0 ? ft.query(r - 1) : 0;
            ft.update(r, 1);
        }
        
        return count;
    }
    
    // 10. 2D Range Sum Query
    static class NumMatrix {
        int[][] matrix;
        int m, n;
        
        public NumMatrix(int[][] matrix) {
            if (matrix.length == 0 || matrix[0].length == 0) return;
            this.matrix = matrix;
            this.m = matrix.length;
            this.n = matrix[0].length;
        }
        
        public void update(int row, int col, int val) {
            matrix[row][col] = val;
        }
        
        public int sumRegion(int row1, int col1, int row2, int col2) {
            int sum = 0;
            for (int i = row1; i <= row2; i++) {
                for (int j = col1; j <= col2; j++) {
                    sum += matrix[i][j];
                }
            }
            return sum;
        }
    }
    
    public static void main(String[] args) {
        FenwickTreeExamples examples = new FenwickTreeExamples();
        
        System.out.println("=== Fenwick Tree Examples ===");
        
        // Test 1: Range Sum Query
        int[] nums = {1, 3, 5, 7, 9, 11};
        NumArray numArray = new NumArray(nums);
        System.out.println("1. Sum Range [1,3]: " + numArray.sumRange(1, 3));
        numArray.update(1, 10);
        System.out.println("   Sum Range [1,3] after update: " + numArray.sumRange(1, 3));
        
        // Test 2: Count Smaller Numbers
        int[] nums2 = {5, 2, 6, 1};
        System.out.println("2. Count Smaller: " + examples.countSmaller(nums2));
        
        // Test 3: Count Range Sum
        int[] nums3 = {-2, 5, -1};
        System.out.println("3. Count Range Sum: " + examples.countRangeSum(nums3, -2, 2));
        
        // Test 4: Reverse Pairs
        int[] nums4 = {1, 3, 2, 3, 1};
        System.out.println("4. Reverse Pairs: " + examples.reversePairs(nums4));
        
        // Test 5: Count Teams
        int[] rating = {2, 5, 3, 4, 1};
        System.out.println("5. Count Teams: " + examples.numTeamsInArray(rating));
        
        // Test 6: Process Queries
        int[] queries = {3, 1, 2, 1};
        System.out.println("6. Process Queries: " + Arrays.toString(examples.processQueries(queries, 5)));
        
        // Test 7: Create Sorted Array
        int[] instructions = {1, 5, 6, 2};
        System.out.println("7. Create Sorted Array Cost: " + examples.createSortedArray(instructions));
        
        // Test 8: Range Add Query
        int[][] updates = {{1, 3, 2}, {2, 4, 3}, {0, 2, -2}};
        System.out.println("8. Range Add Query: " + Arrays.toString(examples.rangeAddQuery(5, updates)));
        
        // Test 9: Count Inversions
        int[] nums9 = {8, 4, 2, 1};
        System.out.println("9. Count Inversions: " + examples.countInversions(nums9));
        
        // Test 10: 2D Range Sum
        int[][] matrix = {{3, 0, 1, 4, 2}, {5, 6, 3, 2, 1}, {1, 2, 0, 1, 5}};
        NumMatrix numMatrix = new NumMatrix(matrix);
        System.out.println("10. 2D Sum Region [1,1,2,3]: " + numMatrix.sumRegion(1, 1, 2, 3));
    }
}